imap_processing/hi/l1a/science_direct_event.py

```diff
@@ -0,0 +1,288 @@
+            (start_bitmask_data, tof_1, tof_2, tof_3, de_tag),
+            .....
+        ]
+        second packet = [
+            (start_bitmask_data, tof_1, tof_2, tof_3, de_tag),
+            .....
+        ]
+
+    In direct event data, if no hit is registered, the tof_x field in
+    the DE to a value of negative one. However, since the field is described as a
+    "10-bit unsigned counter," it cannot actually store negative numbers.
+    Instead, the value negative is represented by the maximum value that can
+    be stored in a 10-bit unsigned integer, which is 0x3FF (in hexadecimal)
+    or 1023 in decimal. This value is used as a special marker to
+    indicate that no hit was registered. Ideally, the system should
+    not be able to register a hit with a value of 1023 for all
+    tof_1, tof_2, tof_3, because this is in case of an error. But,
+    IMAP-Hi like to process it still to investigate the data.
+    Example of what it will look like if no hit was registered.
+            (start_bitmask_data, 1023, 1023, 1023, de_tag)
+            start_bitmask_data will be 1 or 2 or 3.
+
+    Parameters
+    ----------
+    event_data : str
+        48-bits Event data
+
+    Returns
+    -------
+    list
+        Parsed event data
+    """
+    if int(event_data[:2]) == 0:
+        # parse metaevent
+        metaevent = {
+            "start_bitmask_data": int(event_data[:2], 2),
+            "esa_step": int(event_data[2:6], 2),
+            "subseconds": int(event_data[6:16], 2),
+            "seconds": int(event_data[16:], 2),
+        }
+        return metaevent
+
+    # parse direct event
+    direct_event = {
+        "start_bitmask_data": int(event_data[:2], 2),
+        "tof_1": int(event_data[2:12], 2),
+        "tof_2": int(event_data[12:22], 2),
+        "tof_3": int(event_data[22:32], 2),
+        "de_tag": int(event_data[32:], 2),
+    }
+    return direct_event
+
+
+def break_into_bits_size(binary_data: str):
+    """Break binary stream data into 48-bits.
+
+    Parameters
+    ----------
+    binary_data : str
+        Binary data
+
+    Returns
+    -------
+    list
+        List of 48-bits
+    """
+    # TODO: ask Paul what to do if the length of
+    # binary_data is not a multiple of 48
+    return [binary_data[i : i + 48] for i in range(0, len(binary_data), 48)]
+
+
+def create_dataset(de_data_list: list):
+    """Create xarray dataset.
+
+    Parameters
+    ----------
+    de_data_list : list
+        Parsed direct event data list
+
+    Returns
+    -------
+    xarray.Dataset
+        xarray dataset
+    """
+    # These are the variables that we will store in the dataset
+    met_subseconds = []
+    met_seconds = []
+    esa_step = []
+    trigger_id = []
+    tof_1 = []
+    tof_2 = []
+    tof_3 = []
+    de_tag = []
+    for event in de_data_list:
+        # TODO: how to check for fist event?
+        # metaevent.
+        # How to handle if first event is not
+        # metaevent? This means that current
```

Comment:
Let me know. Hi definitely has metaevent. Ultra and hi may be do things differently. metaevent is nothing but a way to store information about bigger portion of time information. Eg. metaevent stores information about, let's say "20240319T09:30:01". Then direct event time tag stores information of time ticks since that time. Then we use those two to combine and get exact time information of each event. I hope that helps.

Reply:
I haven't heard the term metaevent for Lo and the only time field apart from SHCOARSE is the DE time, so I don't think Lo has that either